pkg/calculator/experimentcalc/experiment_calculator.go

```diff
@@ -0,0 +1,650 @@
+	"github.com/bucketeer-io/bucketeer/pkg/calculator/stan"
+	v2es "github.com/bucketeer-io/bucketeer/pkg/calculator/storage/v2"
+	envclient "github.com/bucketeer-io/bucketeer/pkg/environment/client"
+	ecclient "github.com/bucketeer-io/bucketeer/pkg/eventcounter/client"
+	experimentclient "github.com/bucketeer-io/bucketeer/pkg/experiment/client"
+	"github.com/bucketeer-io/bucketeer/pkg/log"
+	"github.com/bucketeer-io/bucketeer/pkg/metrics"
+	"github.com/bucketeer-io/bucketeer/pkg/storage/v2/mysql"
+	"github.com/bucketeer-io/bucketeer/proto/calculator"
+	"github.com/bucketeer-io/bucketeer/proto/environment"
+	"github.com/bucketeer-io/bucketeer/proto/eventcounter"
+	"github.com/bucketeer-io/bucketeer/proto/experiment"
+)
+
+var (
+	failedToSampleErr = errors.New("failed to get all samples")
+)
+
+const (
+	day         = 24 * 60 * 60
+	numOfChains = 5
+)
+
+type ExperimentCalculator struct {
+	httpStan *stan.Stan
+	modelID  string
+
+	environmentClient  envclient.Client
+	eventCounterClient ecclient.Client
+	experimentClient   experimentclient.Client
+	mysqlClient        mysql.Client
+	metrics            metrics.Registerer
+
+	logger *zap.Logger
+}
+
+func NewExperimentCalculator(
+	httpStan *stan.Stan,
+	environmentClient envclient.Client,
+	eventCounterClient ecclient.Client,
+	experimentClient experimentclient.Client,
+	mysqlClient mysql.Client,
+	metrics metrics.Registerer,
+	logger *zap.Logger,
+) *ExperimentCalculator {
+	registerMetrics(metrics)
+	compiledModel, err := httpStan.CompileModel(context.TODO(), stan.ModelCode())
+	if err != nil {
+		logger.Error("Failed to compile model",
+			zap.Error(err),
+		)
+		return nil
+	}
+	modelID := compiledModel.Name[len("models/"):]
+	return &ExperimentCalculator{
+		httpStan: httpStan,
+		modelID:  modelID,
+
+		environmentClient:  environmentClient,
+		eventCounterClient: eventCounterClient,
+		experimentClient:   experimentClient,
+		mysqlClient:        mysqlClient,
+		metrics:            metrics,
+
+		logger: logger.Named("experiment-calculator"),
+	}
+}
+
+func (e ExperimentCalculator) Run(ctx context.Context, request *calculator.BatchCalcRequest) {
+	now := time.Now()
+	// Step 1: Get all the environments
+	environments, environmentErr := e.listEnvironments(ctx)
+	if environmentErr != nil {
+		e.logger.Error("ExperimentCalculator failed to list environments",
+			log.FieldsFromImcomingContext(ctx).AddFields(
+				zap.Error(environmentErr),
+			)...,
+		)
+		return
+	}
+	// Step 2: Get all the events for the experiment
+	for _, env := range environments {
+		experiments, experimentErr := e.listExperiments(ctx, env.Namespace)
+		if experimentErr != nil {
+			e.logger.Error("ExperimentCalculator failed to list experiments",
+				log.FieldsFromImcomingContext(ctx).AddFields(
+					zap.String("namespace", env.Namespace),
+					zap.Error(experimentErr),
+				)...,
+			)
+			return
+		}
+		for _, ex := range experiments {
+			if ex.StartAt > now.Unix() {
+				// already calculated
+				continue
```

Comment:
I think you skip it because the experiment has been started yet, not because it has been already calculated.

Reply:
When wrote this I was also confuse about the logic in Python Calculator, so I just keep it the same, maybe the orginal logic is not right :) :
[the logic in `run` function](https://github.com/bucketeer-io/bucketeer/blob/main/python/src/lib/calculator/job/calculate_experiments.py#L78)

Jump in:
[is_calculated](https://github.com/bucketeer-io/bucketeer/blob/main/python/src/lib/calculator/domain/experiment.py#L41)

Reply:
Right, I think so too.
I also confirmed with @cre8ivejp. 
Please fix it 🙏

Reply:
Thanks for your confirmation!

So this will be changed to:
```go
for _, ex := range experiments {
			if ex.StartAt <= now.Unix() {
				// experiment not started yet 
				continue
```

Reply:
Fixed.